### DISCUSSION TITLE
How to do non-recursive atomic rule.

### DISCUSSION THREAD

**Discussion by lefuturiste** [pest-parser/pest#720] [Q&A]
I'm trying to create a grammar to have key-value data language
The goal is to have key value pairs separated by at least one newline. But I don't want to go in atomic mode because then it's recursive and I have to explicitly specify every whitespaces I want in my key value pair and in all the different sort of values that I can have.

Ideally I want to declare that a grammar that parse this kind of sample: `Key : value` sequence separated with a newline
```
key1:13
key2:549
```
but also allow whitespace between key and value
```
key1: 13
key2: 549
```

This is the grammar that I use
```
WHITESPACE = _{ " " | "\t" | "\r" | "\n" }

key = { (ASCII_ALPHANUMERIC | "_")+ }
int = { ASCII_DIGIT+ }
value = { int | "nil" }
entry = { key ~ ":" ~ value }
entries = ${ entry ~ (WHITESPACE+ ~ entry)* }
```

I could use : `entry = { key ~ WHITESPACE* ~ ":" ~ WHITESPACE* ~ value }` but my final goal is to make a grammar much more complicated so the value in the pair could have many more language trick and specific grammar that I don't want to specify eachtime the whitespaces.

How could I have a whitespace explicit mode (atomic) only on one level and non recursive?

**Comment by tomtau** [comment]
How about removing `\n` from the implicit `WHITESPACE` rule and have one of the rules (e.g. `entry` or `entries`) explicitly have `\n`?

**Reply by lefuturiste** [reply]
Yes but in the future I want my entry value to also have implicit newlines. For example I want to allow also:

```
key: 1243
key: "Value"

key: {
     "hello"
   
     "wow"
     "sooo much newlines"
     @OtherRule {
         "hey"

         "foo"
     }
}
